File: sources/scanners/CitraSource.js
const config = require("../utils/configFormats.js");
const { EmulationSource } = require("./EmulationSource.js");
const common = require("./common.js");
const fsp = require("fs/promises");
const path = require("path");
const { CitraGame } = require("../games/CitraGame.js");

const USER_DIR = process.env["HOME"];
const CITRA_CONFIG_PATH = `${USER_DIR}/.config/citra-emu/qt-config.ini`;
const GAME_FILES_REGEX = /.+\.(3ds|cci)/i;

/**
 * A class representing a Citra source
 */
class CitraSource extends EmulationSource {

	preferCache = false;

	constructor(preferCache = false) {
		super();
		this.preferCache = preferCache;
	}

	/**
	 * Get the citra config data from $HOME/.config/citra-emu/qt-config.ini
	 * Found in $HOME/.config/citra-emu/qt-config.ini
	 * Validates the config data before returning it.
	 * @returns {object} - An object containing citra's config
	 * @private
	 */
	async _getConfig() {

		const configFileContents = await fsp.readFile(CITRA_CONFIG_PATH, "utf-8");
		const configData = config.config2js(configFileContents);

		// Check "UI > Paths\Gamedirs\size" value in config to be numeric
		const nDirs = parseInt(configData["UI"]["Paths\\gamedirs\\size"]);
		if (Number.isNaN(nDirs)) {
			throw Error("Non numeric Paths\\gamedirs\\size value in config file");
		}

		return configData;
	}

	/**
	 * Get citra's game dirs from its config data
	 * @param {object} configData - Citra's config data
	 * @returns {GameDir[]} - The game dirs extracted from citra's config
	 * @private
	 */
	async _getROMDirs(configData) {

		const dirs = [];

		// Get number of paths
		if (typeof configData["UI"] === "undefined") { return dirs; }
		const nDirs = parseInt(configData["UI"]["Paths\\gamedirs\\size"]);

		// Get paths
		for (let i = 1; i <= nDirs; i++) {
			const recursive = String(configData["UI"][`Paths\\gamedirs\\${i}\\deep_scan`]).toLowerCase() === "true";
			const path = configData["UI"][`Paths\\gamedirs\\${i}\\path`];
			if (typeof path === "undefined") { continue; }
			dirs.push(new common.GameDir(path, recursive));
		}

		return dirs;

	}

	/**
	 * Get citra ROM games from given game directories
	 * @param {GameDir[]} dirs - The directories in which to search for ROMs
	 * @returns {CitraGame[]} - An array of found games
	 * @private
	 */
	async _getROMGames(dirs) {

		const gamePaths = await this._getROMs(dirs, GAME_FILES_REGEX);
		const games = [];
		for (const gamePath of gamePaths) {
			const game = new CitraGame(path.basename(gamePath), gamePath);
			game.isInstalled = true;
			games.push(game);
		}
		return games;

	}

	/**
	 * Get citra installed games.
	 * @throws {NotImplementedError} - This is not yet supported
	 * @param {object} config - Citra's config data
	 * @private
	 * @todo
	 */
	async _getInstalledGames(configData) {

		// TODO implement scanning for installed games
		throw new common.NotImplementedError("Scanning for installed citra games is not implemented");

	}

	/**
	 * Get all citra games
	 * @param {boolean} warn - Whether to display additional warnings
	 * @returns {CitraGame[]} - An array of found games
	 */
	async scan(warn = false) {

		// Get config
		let configData;
		try {
			configData = await this._getConfig();
		} catch (error) {
			if (warn){
				console.warn(`Unable to read citra config file : ${error}`);
			}
		}

		// Get ROM dirs
		let romDirs = [];
		if (typeof configData !== "undefined") {
			try {
				romDirs = await this._getROMDirs(configData);
			} catch (error) {
				if (warn){
					console.warn(`Unable to get citra ROM dirs : ${error}`);
				}
			}
		}

		// Get ROM games
		let romGames = [];
		if (romDirs.length > 0) {
			try {
				romGames = await this._getROMGames(romDirs);
			} catch (error) {
				if (warn){
					console.warn(`Unable to get citra ROMs : ${error}`);
				}
			}
		}

		// Get installed games
		let installedGames = [];
		if (typeof configData !== "undefined") {
			try {
				installedGames = await this._getInstalledGames(configData);
			} catch (error) {
				if (warn) {
					console.warn(`Unable to get citra installed games : ${error}`);
				}
			}
		}

		return [...romGames, ...installedGames];

	}

}

module.exports = {
	CitraSource
};

File: gali/sources/citra_source.py
from math import inf
from pathlib import PurePath
from configparser import ConfigParser

from gali.sources.emulation_source import EmulationSource
from gali.utils.locations import USER_DIR
from gali.sources.game_dir import GameDir
from gali.games.citra_game import CitraGame, CitraFlatpakGame

class CitraSource(EmulationSource):

	name           : str             = "Citra"
	game_class     : type[CitraGame] = CitraGame
	config_path    : str             = f"{USER_DIR}/.config/citra-emu/qt-config.ini"
	rom_extensions : tuple[str]      = (".3ds", ".cci")

	def get_config(self) -> dict:
		config = ConfigParser()
		try:
			with open(self.config_path) as file:
				config.read_file(file)
		except IOError as err:
			raise err
		return config

	def get_rom_dirs(self, config) -> list[GameDir]:
		rom_dirs = []
		n_dirs = config.getint("UI", r"Paths\gamedirs\size", fallback=0)
		for i in range(1, n_dirs + 1):
			deep = config.getboolean("UI", fr"Paths\gamedirs\{i}\deep_scan", fallback=False)
			path = config.get("UI", fr"Paths\gamedirs\{i}\path", fallback=None)
			if path is None:
				continue
			if path in ("INSTALLED", "SYSTEM"):
				continue
			depth = inf if deep else 0
			rom_dirs.append(GameDir(path, depth))
		return rom_dirs

	def get_rom_games(self, rom_dirs) -> list[CitraGame]:
		games = []
		for rom_dir in rom_dirs:
			rom_paths = []
			try:
				rom_paths = self.get_rom_paths(rom_dir, self.rom_extensions)
			except Exception as err:
				rom_paths = []
			for path in rom_paths:
				name = PurePath(path).name
				game = self.game_class(
					name=name, 
					game_path=path,
					is_installed=True,
				)
				games.append(game)
		return games

	def scan(self) -> list[CitraGame]:
		config = self.get_config()
		rom_dirs = self.get_rom_dirs(config)
		rom_games = self.get_rom_games(rom_dirs)
		return rom_games

class CitraFlatpakSource(CitraSource):

	name           : str             = "Citra (Flatpak)"
	game_class     : type[CitraGame] = CitraFlatpakGame
	config_path    : str             = f"{USER_DIR}/.var/app/org.citra_emu.citra/config/citra-emu/qt-config.ini"